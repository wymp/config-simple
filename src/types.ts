/** A special symbol indicating that the given value is required */
export const REQUIRED = Symbol.for("required");

/** A type indicating an error. You can return this from a custom transformer function as well */
export type ConfigError = `::ERROR::${string}`;

/**
 * Any function conforming to this interface may be passed to the `configValue` function to transform the value of a given
 * config key.
 *
 * @example
 * ```ts
 * const myType = ((val: string | undefined, varname: string) => {
 *   if (!val) {
 *     if (def === REQUIRED) {
 *       return `::ERROR::${varname}::This value is required`;
 *     }
 *     return def;
 *   }
 *   try {
 *     // Normally you would validate but for testing we're skipping that and just returning whatever
 *     return JSON.parse(val) as MyType;
 *   } catch (e: any) {
 *     return `::ERROR::${varname}::Invalid JSON string passed: ${e.message}`;
 *   }
 * }) as TransformerFunction<MyType>;
 *
 * // ...
 *
 * const configDef = {
 *   myType: configValue('MY_TYPE', myType, REQUIRED),
 *   // ...
 * }
 * ```
 */
export type TransformerFunc<T> = (val: string | undefined, varname: string) => T | undefined | ConfigError;

/**
 * A transformer. See {@link TransformerFunc} for more details.
 */
export type Transformer<T> = { transform: TransformerFunc<T> };

/**
 * To implement a custom validator, just implement a function that takes the value to be validated and returns either
 * undefined or an error of error strings.
 */
export type ValidatorFunc<T> = (val: T | undefined) => void | string | string[];
