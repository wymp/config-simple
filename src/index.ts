export * from "./configValue";
export * from "./transformers/transformers";
export * from "./types";
export * from "./validate";
export * from "./validators/validators";
