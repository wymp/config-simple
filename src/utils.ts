import { ConfigError } from "./types";

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const isConfigError = (val: any): val is ConfigError =>
  Boolean(val && typeof val === "string" && val.startsWith("::ERROR::"));
