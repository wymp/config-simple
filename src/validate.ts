import { ConfigError } from "./types";

type CleanFrozenConfig<T> = T extends object
  ? { readonly [K in keyof T]: CleanFrozenConfig<T[K]> }
  : Exclude<T, ConfigError>;

type ConfigResult<T> =
  | { t: "success"; value: CleanFrozenConfig<T> }
  | {
      t: "error";
      errors: string[];
      value: CleanFrozenConfig<T>;
    };

/**
 * Validate a configuration object and return a frozen copy.
 *
 * @example
 * ```ts
 * export const config = validate(configDef);
 * export const dirtyConfig = validate(configDef, 'dont-throw');
 * ```
 *
 * In the first form, this function will throw an error if the configuration is invalid. In the second form, it will
 * return a `ConfigResult` object that can be inspected for errors.
 *
 * NOTE: THE 'ERROR' SIDE OF THE RESPONSE FROM THE SECOND FORM IS DANGEROUS. We want the option to not throw on config
 * errors, but we also don't want to have to use type-guards to handle bad config throughout our entire codebase. This
 * type is a dangerous compromise that allows us to blindly use values as if they were valid, while knowing that they
 * may not be.
 */
export function validate<T>(obj: T): CleanFrozenConfig<T>;
export function validate<T>(obj: T, behavior: "dont-throw"): ConfigResult<T>;
export function validate<T>(obj: T, behavior?: "dont-throw"): CleanFrozenConfig<T> | ConfigResult<T> {
  const result = reportConfigResult(obj);
  if (behavior === "dont-throw") {
    return result;
  }
  if (result.t === "error") {
    throw new Error(`Invalid configuration: \n\n  * ${result.errors.join("\n  * ")}`);
  }
  return result.value;
}

const reportConfigResult = <T>(obj: T, path: string[] = []): ConfigResult<T> => {
  const errors: string[] = [];
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const response: any = Array.isArray(obj) ? [...obj] : { ...obj };
  for (const k in response) {
    if (Object.prototype.hasOwnProperty.call(response, k)) {
      const val = response[k];
      if (typeof val === "string" && val.startsWith("::ERROR::")) {
        const match = val.match(/^::ERROR::(?:([^:]+)::)?(.*)$/);
        const key = [...path, k].join(".");
        const varname = match?.[1] ? ` (${match[1]})` : "";
        const messages = match?.[2] ? match[2].split("|") : ["Invalid value (unknown error)"];
        if (messages.length === 1) {
          errors.push(`${key}${varname}: ${messages[0]}`);
        } else {
          errors.push(`${key}${varname}:\n    * ${messages.join("\n    * ")}`);
        }
      } else {
        if (response[k] && typeof response[k] === "object") {
          const result = reportConfigResult(response[k], [...path, k]);
          if (result.t === "error") {
            errors.push(...result.errors);
          }
          // eslint-disable-next-line @typescript-eslint/no-explicit-any
          response[k] = result.value as any;
        }
      }
    }
  }

  if (errors.length > 0) {
    return { t: "error", errors, value: Object.freeze(response) as CleanFrozenConfig<T> };
  } else {
    return { t: "success", value: Object.freeze(response) as CleanFrozenConfig<T> };
  }
};
