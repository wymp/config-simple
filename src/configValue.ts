import { Transformers } from "./transformers/transformers";
import { ConfigError, REQUIRED, Transformer, ValidatorFunc } from "./types";
import { isConfigError } from "./utils";
import { required as requiredValidator } from "./validators/required";

type ValidatorArg<T> = ValidatorFunc<T> | ValidatorFunc<T>[];
type DefOrReq<T> = T | typeof REQUIRED;

/** @see {@link configValue} */
export interface ConfigValueFunc {
  (envVarName: string | string[]): string | undefined;
  (envVarName: string | string[], defaultOrRequired: DefOrReq<string>): string | ConfigError;
  (envVarName: string | string[], validator: ValidatorArg<string>): string | undefined | ConfigError;
  (envVarName: string | string[], required: typeof REQUIRED, validator: ValidatorArg<string>): string | ConfigError;
  (envVarName: string | string[], defaultValue: string, validator: ValidatorArg<string>): string;
  (envVarName: string | string[], t: "str"): string | undefined;
  (envVarName: string | string[], t: "str", defaultOrRequired: DefOrReq<string>): string | ConfigError;
  (envVarName: string | string[], t: "str", validator: ValidatorArg<string>): string | undefined | ConfigError;
  (
    envVarName: string | string[],
    t: "str",
    required: typeof REQUIRED,
    validator: ValidatorArg<string>
  ): string | ConfigError;
  (
    envVarName: string | string[],
    t: "str",
    defaultValue: string,
    validator: ValidatorArg<string>
  ): string | ConfigError;
  (envVarName: string | string[], t: "num"): number | undefined | ConfigError;
  (envVarName: string | string[], t: "num", defaultOrRequired: DefOrReq<number>): number | ConfigError;
  (envVarName: string | string[], t: "num", validator: ValidatorArg<number>): number | ConfigError;
  (
    envVarName: string | string[],
    t: "num",
    required: typeof REQUIRED,
    validator: ValidatorArg<number>
  ): number | ConfigError;
  (
    envVarName: string | string[],
    t: "num",
    defaultValue: number,
    validator: ValidatorArg<number>
  ): number | ConfigError;
  (envVarName: string | string[], t: "bool"): boolean | undefined | ConfigError;
  (envVarName: string | string[], t: "bool", defaultOrRequired: DefOrReq<boolean>): boolean | ConfigError;
  (envVarName: string | string[], t: "bool", validator: ValidatorArg<boolean>): boolean | undefined | ConfigError;
  (
    envVarName: string | string[],
    t: "bool",
    required: typeof REQUIRED,
    validator: ValidatorArg<boolean>
  ): boolean | ConfigError;
  (
    envVarName: string | string[],
    t: "bool",
    defaultValue: boolean,
    validator: ValidatorArg<boolean>
  ): boolean | ConfigError;
  <T>(envVarName: string | string[], t: Transformer<T>): T | undefined | ConfigError;
  <T>(envVarName: string | string[], t: Transformer<T>, defaultOrRequired: DefOrReq<T>): T | ConfigError;
  <T>(envVarName: string | string[], t: Transformer<T>, validator: ValidatorArg<T>): T | undefined | ConfigError;
  <T>(
    envVarName: string | string[],
    t: Transformer<T>,
    required: typeof REQUIRED,
    validator: ValidatorArg<T>
  ): T | ConfigError;
  <T>(envVarName: string | string[], t: Transformer<T>, defaultValue: T, validator: ValidatorArg<T>): T | ConfigError;
}

/**
 * Get a value with the given specification. The standard transformers available are 'str', 'num' and 'bool'. You can
 * also pass a custom transformer function (see {@link TransformerFunc}) as well a validator function or array of
 * validator functions. (Import the {@link Validators | `Validators`} library from
 * this library to explore the bundled validators or write your own by implementing {@link ValidatorFunc}.)
 *
 * The most common usage of this will be the following (see more examples below for advanced usage in context):
 *
 * ```ts
 * // Get a possibly-undefined string from the SOME_VAR environment variable
 * configValue('SOME_VAR');
 *
 * // Get a string with a default value of 'some default' from the SOME_VAR environment variable
 * configValue('SOME_VAR', 'some default');
 *
 * // Get a possibly-undefined number from the SOME_VAR environment variable
 * configValue('SOME_VAR', 'num');
 *
 * // Get a number with a default value of 5 from SOME_VAR
 * configValue('SOME_VAR', 'num', 5);
 *
 * // Get a required string from SOME_VAR with no default (note: `REQUIRED` is a special imported symbol from this
 * // library)
 * configValue('SOME_VAR', REQUIRED);
 *
 * // Get a required number from SOME_VAR with no default
 * configValue('SOME_VAR', 'num', REQUIRED);
 * ```
 *
 * @example
 * Following is a more real-world example of how this might all be used in context
 *
 * ```ts
 * const configDef = {
 *   port: configValue('PORT', 'num', 3000),
 *   hiPort: configValue('PORT', 'num', 30_000, hiport),
 *   host: configValue('HOST', 'str', 'http://localhost', hostValidator),
 *   db: {
 *     url: configValue('DB_URL', 'str', 'postgres://postgres@localhost:5432'),
 *     migrateOnStart: configValue('DB_MIGRATE_ON_START', 'bool', false),
 *   },
 *   signatureSecret: configValue('SIGNATURE_SECRET', 'str', [required, exactLength(32), hex]),
 *   secondsToWaitForThing: configValue('SECONDS_TO_WAIT_FOR_THING', 'num', required),
 * }
 *
 * export const config = validate(configDef);
 * ```
 *
 * NOTE: This function uses any-casts. This is because the underlying transformer functions are overloaded and require
 * _either_ one argument configuration _or_ the other, which gets messy in a function like this. Since this function is
 * itself properly overloaded, it's safe to use any-casts to solve this here.
 */
export const configValue: ConfigValueFunc = <T>(
  envVarNameArr: string | string[],
  defOrReqOrTransOrVal?: string | typeof REQUIRED | "str" | "num" | "bool" | Transformer<T> | ValidatorArg<T>,
  defOrReqOrValidator?: ValidatorArg<T> | T | typeof REQUIRED,
  validatorArg?: ValidatorArg<T>
) => {
  const { defaultValue, envVarName, required, transformer, validators } = processArgs(
    envVarNameArr,
    defOrReqOrTransOrVal,
    defOrReqOrValidator,
    validatorArg
  );

  // If required, add our special `required` validator to the validators stack
  if (required) {
    validators.unshift(requiredValidator);
  }

  // Get and transform the value
  let val = transformer(process.env[envVarName], envVarName) as T | undefined | ConfigError;

  // If we don't have a value but do have a default, set our value to that
  if (val === undefined && defaultValue !== undefined) {
    val = defaultValue as T;
  }

  // If the transformation worked, pass the value through our validators
  if (!isConfigError(val) && validators?.length) {
    const errors: string[] = [];
    for (const validator of validators) {
      const newErrors = validator(val);
      if (newErrors) {
        errors.push(...(Array.isArray(newErrors) ? newErrors : [newErrors]));
      }
    }

    if (errors.length > 0) {
      val = `::ERROR::${envVarName}::${errors.join("|")}`;
    }
  }

  // Return whatever we've got
  return val;
};

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const isCustomTransformer = <T>(obj: any): obj is Transformer<T> =>
  obj && typeof obj === "object" && typeof obj.transform === "function";
// eslint-disable-next-line @typescript-eslint/no-explicit-any
const isValidatorArg = <T>(obj: any): obj is ValidatorArg<T> => typeof obj === "function" || Array.isArray(obj);

/**
 * Do some super-intense argument analysis to figure out what's what here and return a normalized set of arguments.
 *
 * God help ye who wanders here............
 */
const processArgs = <T>(
  envVarNameArr: string | string[],
  defOrReqOrTransOrVal?: string | typeof REQUIRED | "str" | "num" | "bool" | Transformer<T> | ValidatorArg<T>,
  defOrReqOrValidator?: ValidatorArg<T> | T | typeof REQUIRED,
  validatorArg?: ValidatorArg<T>
) => {
  // Get the best env var to use
  const envVarName = Array.isArray(envVarNameArr)
    ? envVarNameArr.find((v) => process.env[v] !== undefined) || envVarNameArr[0]!
    : envVarNameArr;

  const required = defOrReqOrTransOrVal === REQUIRED || defOrReqOrValidator === REQUIRED;

  // Get the transformer function to use
  const transformer = isCustomTransformer<T>(defOrReqOrTransOrVal)
    ? defOrReqOrTransOrVal.transform
    : typeof defOrReqOrTransOrVal === "string" && defOrReqOrTransOrVal in Transformers
      ? Transformers[defOrReqOrTransOrVal as keyof typeof Transformers].transform
      : Transformers.str.transform;

  // Get and normalize validators, if there are any
  const validators = validatorArg
    ? Array.isArray(validatorArg)
      ? validatorArg
      : [validatorArg]
    : isValidatorArg<T>(defOrReqOrValidator)
      ? typeof defOrReqOrValidator === "function"
        ? [defOrReqOrValidator]
        : defOrReqOrValidator
      : isValidatorArg<T>(defOrReqOrTransOrVal)
        ? typeof defOrReqOrTransOrVal === "function"
          ? [defOrReqOrTransOrVal]
          : defOrReqOrTransOrVal
        : [];

  // Get the default value, if specified
  const defaultValue =
    defOrReqOrTransOrVal !== REQUIRED &&
    !isValidatorArg(defOrReqOrTransOrVal) &&
    !isCustomTransformer<T>(defOrReqOrTransOrVal) &&
    !((defOrReqOrTransOrVal as string) in Transformers)
      ? (defOrReqOrTransOrVal as T | undefined)
      : defOrReqOrValidator !== undefined && !isValidatorArg(defOrReqOrValidator) && defOrReqOrValidator !== REQUIRED
        ? (defOrReqOrValidator as T)
        : undefined;

  // Return everything all bundled up
  return {
    defaultValue,
    envVarName,
    required,
    transformer,
    validators,
  };
};
