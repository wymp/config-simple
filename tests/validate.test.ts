/* eslint-disable @typescript-eslint/no-explicit-any */
import { ConfigError } from "../src/types";
import { validate } from "../src/validate";

describe("validate", () => {
  test("should return the same object if there are no errors", () => {
    const obj = { a: 1, b: "2", c: { d: 3 } };
    expect(validate(obj)).toStrictEqual(obj);
  });

  test("should throw an error if there are errors", () => {
    const obj = { a: 1, b: "2", c: { d: "::ERROR::bad" }, e: "::ERROR::more bad" };
    expect(() => validate(obj)).toThrowErrorMatchingSnapshot();
  });

  test('should return a ConfigResult if there are errors and behavior is "dont-throw"', () => {
    const obj = {
      a: 1 as number | ConfigError,
      b: "2" as string | undefined | ConfigError,
      c: {
        d: "::ERROR::bad" as boolean | undefined | ConfigError,
      },
    };
    expect(validate(obj, "dont-throw")).toMatchSnapshot();
  });

  test("should return a deeply frozen object", () => {
    const obj = { a: 1, b: "2", c: { d: 3 } };
    const result = validate(obj);
    expect(Object.isFrozen(result)).toBe(true);
    expect(Object.isFrozen(result.c)).toBe(true);
  });

  test("should return both object path and env var name in error", () => {
    const obj = {
      a: 1 as number | ConfigError,
      b: "2" as string | undefined | ConfigError,
      c: {
        d: "::ERROR::MY_ENV_VAR::bad" as boolean | undefined | ConfigError,
      },
    };
    const result = validate(obj, "dont-throw");
    expect((result as any).errors?.[0]).toMatch(/\bc.d\b/);
    expect((result as any).errors?.[0]).toMatch(/\bMY_ENV_VAR\b/);
  });

  test("handles multiple errors per key", () => {
    const obj = { a: "::ERROR::MY_ENV_VAR::bad|worse|worst" };
    expect(() => validate(obj)).toThrowErrorMatchingSnapshot();
  });

  test("should return arrays as arrays", () => {
    const obj = { a: [1, 2, 3] };
    expect(Array.isArray(validate(obj).a)).toBe(true);
  });
});
