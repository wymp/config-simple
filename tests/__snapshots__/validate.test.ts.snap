// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validate handles multiple errors per key 1`] = `
"Invalid configuration: 

  * a (MY_ENV_VAR):
    * bad
    * worse
    * worst"
`;

exports[`validate should return a ConfigResult if there are errors and behavior is "dont-throw" 1`] = `
{
  "errors": [
    "c.d: bad",
  ],
  "t": "error",
  "value": {
    "a": 1,
    "b": "2",
    "c": {
      "d": "::ERROR::bad",
    },
  },
}
`;

exports[`validate should throw an error if there are errors 1`] = `
"Invalid configuration: 

  * c.d: bad
  * e: more bad"
`;
